File: Android/tuilivekit/component/barrage/src/main/java/com/trtc/uikit/component/barrage/view/EmojiLayout.java
package com.trtc.uikit.component.barrage.view;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.tencent.qcloud.tuicore.util.ScreenUtil;
import com.trtc.uikit.component.barrage.R;

import java.util.ArrayList;
import java.util.List;

@SuppressLint("ViewConstructor")
public class EmojiLayout extends FrameLayout {

    private final Context      context;
    private final RecyclerView emojiView;
    private final CardView     deleteView;

    private final int emojiSpace;

    private       EmojiListener emojiListener;
    private final List<Integer> emojiResIds = new ArrayList<>();

    public EmojiLayout(@NonNull Context context, List<Integer> emojiResIds) {
        super(context);
        inflate(context, R.layout.livekit_barrage_emoji_view, this);
        this.context = context;
        this.emojiResIds.addAll(emojiResIds);
        this.emojiSpace = ScreenUtil.dip2px(13);
        emojiView = findViewById(R.id.rv_emoji_list);
        deleteView = findViewById(R.id.cd_delete);
        initViews();
    }

    private void initViews() {
        int padding = emojiSpace / 2;
        emojiView.setPadding(padding, 0, padding, 0);
        int emojiSpanCount = 8;
        GridLayoutManager gridLayoutManager = new GridLayoutManager(context, emojiSpanCount);
        emojiView.setLayoutManager(gridLayoutManager);
        emojiView.setAdapter(new RecyclerView.Adapter<ImageViewHolder>() {
            @NonNull
            @Override
            public ImageViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
                int parentWidth = ((ViewGroup) emojiView.getParent()).getWidth();
                int itemSize = (parentWidth - 2 * (emojiSpanCount + 1) * padding) / emojiSpanCount + 2 * padding;
                ImageView imageView = new ImageView(context);
                ViewGroup.LayoutParams params = new LayoutParams(itemSize, itemSize);
                imageView.setLayoutParams(params);
                return new ImageViewHolder(imageView, padding);
            }

            @Override
            public void onBindViewHolder(@NonNull ImageViewHolder holder, int position) {
                int resId = emojiResIds.get(position);
                holder.imageView.setOnClickListener(v -> {
                    if (emojiListener != null) {
                        emojiListener.onAddEmoji(resId);
                    }
                });
                holder.bind(resId);
            }

            @Override
            public int getItemCount() {
                return emojiResIds.size();
            }
        });
        deleteView.setOnClickListener(v -> {
            if (emojiListener != null) {
                emojiListener.onDelete();
            }
        });
    }

    public void setDeleteViewEnable(boolean enable) {
        deleteView.getChildAt(0).setEnabled(enable);
    }

    public void setEmojiListener(EmojiListener listener) {
        emojiListener = listener;
    }

    private static class ImageViewHolder extends RecyclerView.ViewHolder {
        private final ImageView imageView;
        private final int       padding;

        public ImageViewHolder(@NonNull View itemView, int padding) {
            super(itemView);
            this.imageView = (ImageView) itemView;
            this.padding = padding;
        }

        void bind(int resId) {
            Drawable drawable = ResourcesCompat.getDrawable(itemView.getResources(), resId, null);
            imageView.setPadding(padding, padding, padding, padding);
            imageView.setImageDrawable(drawable);
        }
    }

    public interface EmojiListener {
        void onDelete();

        void onAddEmoji(int resId);
    }

}

File: Android/tuilivekit/src/main/java/com/trtc/uikit/livekit/component/barrage/view/EmojiLayout.kt
package com.trtc.uikit.livekit.component.barrage.view

import android.annotation.SuppressLint
import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.cardview.widget.CardView
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.tencent.qcloud.tuicore.util.ScreenUtil
import com.trtc.uikit.livekit.R

@SuppressLint("ViewConstructor")
class EmojiLayout(context: Context, emojiResIds: List<Int>) : FrameLayout(context) {
    private val emojiView: RecyclerView
    private val deleteView: CardView
    private val emojiSpace = ScreenUtil.dip2px(13f)
    private var emojiListener: EmojiListener? = null
    private val emojiResIds = emojiResIds.toMutableList()

    init {
        inflate(context, R.layout.livekit_barrage_emoji_view, this)
        emojiView = findViewById(R.id.rv_emoji_list)
        deleteView = findViewById(R.id.cd_delete)
        initViews()
    }

    private fun initViews() {
        val padding = emojiSpace / 2
        emojiView.setPadding(padding, 0, padding, 0)
        val emojiSpanCount = 8
        emojiView.layoutManager = GridLayoutManager(context, emojiSpanCount)
        emojiView.adapter = object : RecyclerView.Adapter<ImageViewHolder>() {
            override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ImageViewHolder {
                val parentWidth = (emojiView.parent as ViewGroup).width
                val itemSize = (parentWidth - 2 * (emojiSpanCount + 1) * padding) / emojiSpanCount + 2 * padding
                return ImageViewHolder(ImageView(context).apply {
                    layoutParams = LayoutParams(itemSize, itemSize)
                }, padding)
            }

            override fun onBindViewHolder(holder: ImageViewHolder, position: Int) {
                val resId = emojiResIds[position]
                holder.imageView.setOnClickListener {
                    emojiListener?.onAddEmoji(resId)
                }
                holder.bind(resId)
            }

            override fun getItemCount() = emojiResIds.size
        }
        deleteView.setOnClickListener {
            emojiListener?.onDelete()
        }
    }

    fun setDeleteViewEnable(enable: Boolean) {
        deleteView.getChildAt(0).isEnabled = enable
    }

    fun setEmojiListener(listener: EmojiListener) {
        emojiListener = listener
    }

    private class ImageViewHolder(itemView: View, private val padding: Int) : RecyclerView.ViewHolder(itemView) {
        val imageView: ImageView = itemView as ImageView

        fun bind(resId: Int) {
            val drawable = ResourcesCompat.getDrawable(itemView.resources, resId, null)
            imageView.setPadding(padding, padding, padding, padding)
            imageView.setImageDrawable(drawable)
        }
    }

    interface EmojiListener {
        fun onDelete()
        fun onAddEmoji(resId: Int)
    }
}
